tests/components/ring/test_init.py

```diff
@@ -32,3 +38,116 @@ async def test_setup(hass: HomeAssistant, requests_mock: requests_mock.Mocker) -
+    requests_mock: requests_mock.Mocker,
+    mock_config_entry: MockConfigEntry,
+    caplog,
+) -> None:
+    """Test auth failure on setup entry."""
+    mock_config_entry.add_to_hass(hass)
+    with patch(
+        "ring_doorbell.Ring.update_data",
+        side_effect=AuthenticationError,
+    ):
+        assert not any(mock_config_entry.async_get_active_flows(hass, {SOURCE_REAUTH}))
+        await hass.config_entries.async_setup(mock_config_entry.entry_id)
+        await hass.async_block_till_done()
+        assert any(mock_config_entry.async_get_active_flows(hass, {SOURCE_REAUTH}))
+        assert mock_config_entry.state is ConfigEntryState.SETUP_ERROR
+
+
+@pytest.mark.parametrize("update_type", ["global", "device"])
+async def test_auth_failed_on_update(
+    hass: HomeAssistant,
+    requests_mock: requests_mock.Mocker,
+    mock_config_entry: MockConfigEntry,
+    update_type,
+) -> None:
+    """Test auth failure on data update."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+    assert not any(mock_config_entry.async_get_active_flows(hass, {SOURCE_REAUTH}))
+
+    if update_type == "global":
+        patch_method = "ring_doorbell.Ring.update_devices"
+    else:
+        patch_method = "ring_doorbell.RingDoorBell.history"
+
+    with patch(
+        patch_method,
+        side_effect=AuthenticationError,
+    ):
+        if update_type == "global":
+            await hass.data[DOMAIN][mock_config_entry.entry_id][
+                "device_data"
+            ].async_refresh_all()
+
+        else:
+            hass.data[DOMAIN][mock_config_entry.entry_id]["history_data"].refresh_all()
+        await hass.async_block_till_done()
+        assert any(mock_config_entry.async_get_active_flows(hass, {SOURCE_REAUTH}))
+
+
+error_testdata = [
+    ("global", RingTimeout, "WARNING", "Time out fetching Ring device data"),
+    ("global", RingError, "ERROR", "Error fetching Ring device data: "),
+    (
+        "device",
+        RingTimeout,
+        "WARNING",
+        "Time out fetching Ring history data for device aacdef123",
+    ),
+    (
+        "device",
+        RingError,
+        "ERROR",
+        "Error fetching Ring history data for device aacdef123: ",
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    ("update_type", "error_type", "log_name", "log_msg"),
+    error_testdata,
+    ids=["global-timeout", "global-error", "device-timeout", "device-error"],
+)
+async def test_error_on_update(
+    hass: HomeAssistant,
+    requests_mock: requests_mock.Mocker,
+    mock_config_entry: MockConfigEntry,
+    caplog,
+    update_type,
+    error_type,
+    log_name,
+    log_msg,
+) -> None:
+    """Test auth failure on data update."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    if update_type == "global":
```

Comment:
Here and above: The preferred approach is just have the patch method as the parameterized method, avoiding conditionals in tests. (Then If the behavior varies too much, then just copying the test and code is preferred)

Reply:
Fixed